utils/cpu.go

```diff
@@ -92,9 +113,109 @@ func (c *CPUStats) monitorCPULoad() {
 				c.warningThrottle(func() { logger.Infow("high cpu load", "load", 1-idleRatio) })
 			}
 
-			if c.updateCallback != nil {
-				c.updateCallback(idle)
+			if c.idleCallback != nil {
+				c.idleCallback(idle)
 			}
 		}
 	}
 }
+
+func (c *CPUStats) monitorProcCPULoad() {
+	numCPU := c.platform.numCPU()
+
+	fs, err := procfs.NewFS(procfs.DefaultMountPoint)
+	if err != nil {
+		logger.Errorw("failed read proc fs", err)
+		return
+	}
+	totalCPU, err := getPodCPUCount(fs)
+	if err != nil {
+		logger.Errorw("failed to read pod cpu count", err)
+		return
+	}
+
+	self, err := fs.Self()
+	if err != nil {
+		logger.Errorw("failed to read self", err)
+		return
+	}
+
+	ticker := time.NewTicker(time.Second)
+	defer ticker.Stop()
+
+	var prevTotalTime float64
+	var prevStats map[int]procfs.ProcStat
+	for {
+		select {
+		case <-c.closeChan:
+			return
+		case <-ticker.C:
+			nextStats := make(map[int]procfs.ProcStat)
+			procs, err := procfs.AllProcs()
+			if err != nil {
+				logger.Errorw("failed to read processes", err)
+				continue
+			}
+
+			total, err := fs.Stat()
+			if err != nil {
+				logger.Errorw("failed to read stats", err)
+				continue
+			}
+
+			ppids := make(map[int]int)
+			for _, proc := range procs {
+				nextStats[proc.PID], err = proc.Stat()
+				if err != nil {
+					logger.Errorw("failed to read proc stats", err)
+					continue
+				}
+				if proc.PID != self.PID {
+					ppids[proc.PID], err = getPPID(proc.PID)
+					if err != nil {
+						logger.Errorw("failed to get PPID", err)
+						continue
+					}
+				}
+			}
+
+			nextTotalTime := total.CPUTotal.User + total.CPUTotal.Nice + total.CPUTotal.System + total.CPUTotal.Idle
```

Comment:
I can't really judge whether or not that is the proper measure of total time. Could you document/link to the logic behind it?

Reply:
Might be the reason for the discrepancy - according to https://stackoverflow.com/questions/23367857/accurate-calculation-of-cpu-usage-given-in-percentage-in-linux, it's common to only use those 4, but there are more that should go into it